quinn-proto/src/connection/streams/state.rs

```diff
@@ -899,8 +966,16 @@ pub(super) fn get_or_insert_send(
 #[inline]
 pub(super) fn get_or_insert_recv(
     initial_max_data: u64,
-) -> impl Fn(&mut Option<Box<Recv>>) -> &mut Box<Recv> {
-    move |opt| opt.get_or_insert_with(|| Recv::new(initial_max_data))
+) -> impl FnMut(&mut Option<StreamRecv>) -> &mut Recv {
+    move |opt| {
+        *opt = opt.take().map(|s| match s {
+            StreamRecv::Ready(recv) => StreamRecv::Open(recv),
+            s => s,
+        });
+        opt.get_or_insert_with(|| StreamRecv::Open(Recv::new(initial_max_data)))
+            .open_recv_mut()
+            .unwrap()
```

Comment:
Fold these together? Maybe `opt.get_or_insert_with(...).map(|x| x.ensure_open())`.

Reply:
I don't think I can do what you're suggesting: ensure_open() would need to take by value to go from Free to Open, but get_or_insert_with returns a reference. And the closure takes `&mut Option<StreamRecv>` which is why I have the weird take().map()

Reply:
Ah, you're right.